### DISCUSSION TITLE
Unsuccessful MAGIC installation

### DISCUSSION THREAD

**Discussion by miguelsofo** [RTimothyEdwards/magic#198] [General]
Hi, I been trying to install MAGIC in:

Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.5 LTS
Release:	20.04
Codename:	focal

I been following the: ./configure, make, make install steps. 

during make, prints several warnings, and at the end the following:

display.c: In function ‘rl_redisplay’:
display.c:602:15: error: implicit declaration of function ‘wcwidth’ [-Werror=implicit-function-declaration]
  602 |        temp = wcwidth (wc);

The error with make install is:

<command-line>: warning: "CAD_DIR" redefined
<command-line>: note: this is the location of the previous definition
make[2]: *** No rule to make target '../garouter/libgarouter.o', needed by 'magic'.  Stop.
make[2]: *** No rule to make target '../utils/libutils.a', needed by 'net2ir'.  Stop.

Here are the outputs. 

[config.txt](https://github.com/RTimothyEdwards/magic/files/9990322/config.txt)
[make.txt](https://github.com/RTimothyEdwards/magic/files/9990320/make.txt)
[make_install.txt](https://github.com/RTimothyEdwards/magic/files/9990325/make_install.txt)

I also tried with ./configure --disable-readline without succeed. 

Any ideas? 
Thank you!

**Comment by RTimothyEdwards** [comment]
display.c in the readline code needs `#include <wchar.h>`, according to the man page for `wcwidth`.

However, the fact that readline is compiling suggests that you are compiling the non-Tcl/Tk version of magic, which will cause you to lose a huge number of features.  I suggest making sure that you have all the dependencies, in particular the `tcl-dev` and `tk-dev` packages.

**Reply by miguelsofo** [reply]
Thank you! 

I did: 

./configure --with-tcl --disable-readline --with-opengl --with-tk
Make
Sudo make install

and everything runed smoothly.

**Comment by dlmiles** [comment]
The original problem reported here is believed to be properly fixed via this recent commit in the queue https://github.com/RTimothyEdwards/magic/pull/329/commits/4ec7467120d56e31507d8f6f429b7be306e63074

While I understand the OP may have worked around the issue by using `apt-get install libreadline-dev` so the magic configure detects the system libreadline and uses it instead.

Maybe an upgraded readline (current is 8.x)  if we look here:
https://git.savannah.gnu.org/cgit/readline.git/tree/display.c
https://git.savannah.gnu.org/cgit/readline.git/tree/config.h.in

We can see at some point readline was changed due to incorrect use of feature_test_macros(7)
The `#define` was removed and put into the TOP of config.h which is always the first `#include`.
